src/main/java/org/haffson/adventofcode/days/day07/StepSorter.java

```diff
@@ -0,0 +1,56 @@
+package org.haffson.adventofcode.days.day07;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+public class StepSorter {
+    @Nonnull
+    private final SortedSteps sortedSteps;
+
+    public StepSorter() {
+        this.sortedSteps = new SortedSteps();
+    }
+
+    @Nonnull
+    public String getSortedSteps(@Nonnull final List<Step> stepsInput) {
+        requireNonNull(stepsInput, "stepsInput");
+
+        final List<Step> stepsCopy = new ArrayList<>(stepsInput);
+
+        stepsInput.forEach(step -> addNextStepToResult(stepsCopy));
+
+        return sortedSteps.getStepsAsString();
+    }
+
+
+    private void addNextStepToResult(@Nonnull final List<Step> steps) {
+        requireNonNull(steps, "steps");
+
+        final List<Step> availableSteps = getAvailableSteps(steps);
+        final List<Step> sortedAvailableSteps = new ArrayList<>(availableSteps);
+        sortedAvailableSteps.sort(Comparator.comparing(Step::stepName));
+        final Step nextStep = sortedAvailableSteps.get(0);
+        sortedSteps.addStep(nextStep);
+        steps.remove(nextStep);
```

Comment:
Hier steckt die Annahme drin, dass `steps` manipulierbar ist.
Kann man in einer privaten Methode aber natürlich problemlos machen.

Reply:
Genau. Deswegen hatte ich vorher auch eine copyList mit `newArrayList<>(oldList)` erstellt.
Bin mir aber auch noch nich sicher, ob ich das wirklich so ansprechend finde...

Reply:
Die Kopie ist zumindest eine sichere Wahl.